File: crypto_bot/src/features/patterns.py
"""Rule-based candlestick patterns. No ML in v1."""

from __future__ import annotations

import pandas as pd


def add_patterns(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()
    o, h, l, c = out["open"], out["high"], out["low"], out["close"]
    body = (c - o).abs()
    rng = (h - l).clip(lower=1e-12)
    uw = h - pd.concat([o, c], axis=1).max(axis=1)
    lw = pd.concat([o, c], axis=1).min(axis=1) - l
    close_pos = (c - l) / rng
    prev_o, prev_c = o.shift(1), c.shift(1)
    prev_body = (prev_c - prev_o).abs()
    prev_top = pd.concat([prev_o, prev_c], axis=1).max(axis=1)
    prev_bot = pd.concat([prev_o, prev_c], axis=1).min(axis=1)
    top = pd.concat([o, c], axis=1).max(axis=1)
    bot = pd.concat([o, c], axis=1).min(axis=1)

    doji = body / rng <= 0.1
    hammer = (lw >= 2 * body) & (lw >= 0.55 * rng) & (close_pos >= 0.66) & (c >= o)
    star = (uw >= 2 * body) & (uw >= 0.55 * rng) & ((h - c) / rng >= 0.66) & (c <= o)
    bull_eng = (prev_c < prev_o) & (c > o) & (bot <= prev_bot) & (top >= prev_top) & (body > prev_body)
    bear_eng = (prev_c > prev_o) & (c < o) & (bot <= prev_bot) & (top >= prev_top) & (body > prev_body)
    inside = (h < h.shift(1)) & (l > l.shift(1))

    name = pd.Series("none", index=out.index)
    direction = pd.Series(0, index=out.index, dtype=int)
    strength = pd.Series(0.0, index=out.index)
    name = name.mask(doji, "doji")
    name = name.mask(inside, "inside_bar")
    name = name.mask(star, "shooting_star")
    name = name.mask(hammer, "hammer")
    name = name.mask(bear_eng, "bearish_engulfing")
    name = name.mask(bull_eng, "bullish_engulfing")
    direction = direction.mask(hammer | bull_eng, 1)
    direction = direction.mask(star | bear_eng, -1)
    strength = strength.mask(hammer | star, (lw.clip(lower=uw) / rng).clip(0, 1))
    strength = strength.mask(bull_eng | bear_eng, (body / (prev_body + 1e-12)).clip(0, 1))
    strength = strength.mask(doji, (1 - body / rng).clip(0, 1))
    out["pattern"] = name
    out["pattern_dir"] = direction
    out["pattern_strength"] = strength
    return out

File: src/lib/quant/features/patterns.ts
import type { FeatureBar, PatternHit, PatternName } from "../types";

function body(bar: FeatureBar): number {
  return Math.abs(bar.close - bar.open);
}

function range(bar: FeatureBar): number {
  return Math.max(bar.high - bar.low, 1e-12);
}

function upperWick(bar: FeatureBar): number {
  return bar.high - Math.max(bar.open, bar.close);
}

function lowerWick(bar: FeatureBar): number {
  return Math.min(bar.open, bar.close) - bar.low;
}

function strength01(value: number, lo: number, hi: number): number {
  return Math.min(1, Math.max(0, (value - lo) / (hi - lo)));
}

/**
 * Rule-based candle patterns at bar i using only bars <= i.
 * Returns the strongest pattern for that bar, or null.
 */
export function detectPattern(bars: FeatureBar[], i: number): PatternHit | null {
  if (i < 0 || i >= bars.length) return null;
  const c = bars[i]!;
  const r = range(c);
  const b = body(c);
  const uw = upperWick(c);
  const lw = lowerWick(c);
  const hits: PatternHit[] = [];

  // Doji: tiny body vs range
  if (b / r <= 0.1) {
    hits.push({
      name: "doji",
      direction: 0,
      strength: strength01(1 - b / r, 0.9, 1),
      index: i,
    });
  }

  // Hammer / pin bar bullish: long lower wick, close in upper third
  const closePos = (c.close - c.low) / r;
  if (lw >= 2 * b && lw >= 0.55 * r && closePos >= 0.66 && c.close >= c.open) {
    hits.push({
      name: "hammer",
      direction: 1,
      strength: strength01(lw / r, 0.55, 0.8),
      index: i,
    });
    hits.push({
      name: "pin_bar",
      direction: 1,
      strength: strength01(lw / r, 0.5, 0.78),
      index: i,
    });
  }

  // Shooting star / pin bar bearish
  const closeFromHigh = (c.high - c.close) / r;
  if (uw >= 2 * b && uw >= 0.55 * r && closeFromHigh >= 0.66 && c.close <= c.open) {
    hits.push({
      name: "shooting_star",
      direction: -1,
      strength: strength01(uw / r, 0.55, 0.8),
      index: i,
    });
    hits.push({
      name: "pin_bar",
      direction: -1,
      strength: strength01(uw / r, 0.5, 0.78),
      index: i,
    });
  }

  if (i >= 1) {
    const p = bars[i - 1]!;
    const pb = body(p);
    const prevBull = p.close > p.open;
    const prevBear = p.close < p.open;
    const currBull = c.close > c.open;
    const currBear = c.close < c.open;
    const pTop = Math.max(p.open, p.close);
    const pBot = Math.min(p.open, p.close);
    const cTop = Math.max(c.open, c.close);
    const cBot = Math.min(c.open, c.close);

    if (prevBear && currBull && cBot <= pBot && cTop >= pTop && b > pb) {
      hits.push({
        name: "bullish_engulfing",
        direction: 1,
        strength: strength01(b / (pb + 1e-12), 1, 2.2),
        index: i,
      });
    }
    if (prevBull && currBear && cBot <= pBot && cTop >= pTop && b > pb) {
      hits.push({
        name: "bearish_engulfing",
        direction: -1,
        strength: strength01(b / (pb + 1e-12), 1, 2.2),
        index: i,
      });
    }

    // Inside bar
    if (c.high < p.high && c.low > p.low) {
      hits.push({
        name: "inside_bar",
        direction: 0,
        strength: strength01((p.high - p.low - r) / (p.high - p.low), 0, 0.5),
        index: i,
      });
    }
  }

  if (hits.length === 0) return null;
  hits.sort((a, b2) => b2.strength - a.strength);
  return hits[0]!;
}

export function applyPatterns(bars: FeatureBar[]): PatternHit[] {
  const found: PatternHit[] = [];
  for (let i = 1; i < bars.length; i++) {
    const hit = detectPattern(bars, i);
    bars[i]!.pattern = hit;
    if (hit && hit.strength >= 0.35) found.push(hit);
  }
  return found;
}

export const PATTERN_LABEL: Record<PatternName, string> = {
  pin_bar: "Pin bar",
  hammer: "Hammer",
  shooting_star: "Shooting star",
  bullish_engulfing: "Bullish engulfing",
  bearish_engulfing: "Bearish engulfing",
  inside_bar: "Inside bar",
  doji: "Doji",
};
